# Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [62]:
df.head()
# вивела для себе подивитися датафрейм всіх колонок

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [149]:
df.groupby('Payment').size().sort_values(ascending=False)

Payment
Ewallet        345
Cash           344
Credit card    311
dtype: int64

In [24]:
df.columns
# вивела для себе назви всіх колонок

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

In [26]:
df.groupby('Branch')['gross income'].agg(['sum']) 
# вивожу сумарний прибуток за кожним філіалом

sum
Branch           
A       5057.1605
B       5057.0320
C       5265.1765

In [66]:
profit_branches = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)
# тут вже сортую від найбільшого до найменшого

In [76]:
print("Прибутки філіалів:\n", profit_branches)

Прибутки філіалів:
 Branch
C    5265.1765
A    5057.1605
B    5057.0320
Name: gross income, dtype: float64


In [89]:
top_branch = profit_branches.idxmax()
# передає перше значення з Series

In [91]:
print("Найприбутковіший філіал це:\n", top_branch)

Найприбутковіший філіал це:
 C


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [94]:
city_of_top_branch = df[df['Branch'] == top_branch]['City'].iloc[0]
print("Місто найприбутковішого філіалу це:\n", city_of_top_branch)

Місто найприбутковішого філіалу це:
 Naypyitaw


### 2й варіант

In [97]:
branch_city = df[['Branch', 'City']].drop_duplicates()
city_of_top_branch_2 = branch_city[branch_city['Branch'] == top_branch]['City'].values[0]
print("Місто найприбутковішого філіалу це:\n", city_of_top_branch_2)

Місто найприбутковішого філіалу це:
 Naypyitaw


In [99]:
city_of_top_branch == city_of_top_branch_2

True

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [117]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='nunique')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [133]:
best_elect_branch = invoices_by_category['Electronic accessories'].sort_values(ascending=False)
print("Найбільше інвойсів із покупкою товарів категорії 'Electronic accessories' у філіалі:\n", best_elect_branch.idxmax())

Найбільше інвойсів із покупкою товарів категорії 'Electronic accessories' у філіалі:
 A


In [135]:
best_elect_branch = invoices_by_category['Fashion accessories'].sort_values(ascending=False)
print("Найбільше інвойсів із покупкою товарів категорії 'Fashion accessories' у філіалі:\n", best_elect_branch.idxmax())

Найбільше інвойсів із покупкою товарів категорії 'Fashion accessories' у філіалі:
 C


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


# Гіпотеза 1. 
## Які типи оплат мають найбільші чеки - Payment

In [166]:
df.groupby('Payment')['Total'].mean().sort_values(ascending=False)

Payment
Cash           326.181890
Credit card    324.009878
Ewallet        318.820600
Name: Total, dtype: float64

Висновок: Гіпотеза підтвердилась.
Неочікувано, але найбільший прибуток приносять клієнти, які платять готівкою. 

Рекомендації: 
- Пропонувати знижку/бонуси на інші методи оплати, щоб оптимізувати витрати на обслуговування готівкою. 

# Гіпотеза 2. 
## Коли протягом тижня середній прибуток найвищий

In [160]:
df['Date'] = pd.to_datetime(df['Date'])
df['Day of Week'] = df['Date'].dt.day_name()

In [162]:
df.groupby('Day of Week')['gross income'].mean().sort_values(ascending=False)

Day of Week
Saturday     16.295241
Sunday       15.917613
Thursday     15.648464
Tuesday      15.516047
Friday       15.048421
Wednesday    14.562483
Monday       14.437744
Name: gross income, dtype: float64

Висновок: 
Гіпотеза підтвердилася. У вихідні дні клієнти витрачають більше, тож середній прибуток вищий.
- Найвищий прибуток: Субота (16.30) і Неділя (15.92)
- Найнижчий прибуток: Понеділок (14.44) і Середа (14.56)

Рекомендації:
- Запускати акції на вихідні, щоб ще більше підвищити прибуток.
- У менш прибуткові дні (наприклад, понеділок) варто проводити спеціальні пропозиції, знижки або події, щоб збільшити потік клієнтів.

# Гіпотеза 3.
## Що купують найчастіше залежно від статі?

In [178]:
hyposes_by_gender_category = df.pivot_table(index='Gender', columns='Product line', values='Invoice ID', aggfunc='count')
hyposes_by_gender_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Gender                                                                          
Female                            84                   96                  90   
Male                              86                   82                  84   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Gender                                                                  
Female                       64                  79                 88  
Male                         88                  81                 78

In [186]:
hyposes_by_gender_category.loc['Female'].sort_values(ascending=False).head(3)

Product line
Fashion accessories    96
Food and beverages     90
Sports and travel      88
Name: Female, dtype: int64

In [188]:
hyposes_by_gender_category.loc['Male'].sort_values(ascending=False).head(3)

Product line
Health and beauty         88
Electronic accessories    86
Food and beverages        84
Name: Male, dtype: int64

Висновки:
Я проаналізувала кількість проданих одиниць товарів у різних категоріях залежно від статі покупців.
Жінки більше купують аксесуари та спортивні товари, а чоловіки — електроніку та засоби догляду.
Хочи ми можемо бачити що і в тій, і в іншій групі в топ-3 входить категотія Food and beverages

Результати (топ-3 категорії за кількістю покупок):
- Жінки (Female):
  1. Fashion accessories — 96 одиниць
  2. Food and beverages — 90 одиниць
  3. Sports and travel — 88 одиниць
- Чоловіки (Male):
  1. Health and beauty — 88 одиниць
  2. Electronic accessories — 86 одиниць
  3. Food and beverages — 84 одиниць

Рекомендації:
- Розробити таргетовані маркетингові кампанії, які підкреслюють популярні категорії для кожної статі.
- Формувати персоналізовані рекомендації товарів у інтернет-магазині.
